mpy-cross/wasm.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2023 Damien P. George
+ * Copyright (c) 2023 Jim Mussared
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <emscripten.h>
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "py/builtin.h"
+#include "py/compile.h"
+#include "py/persistentcode.h"
+#include "py/runtime.h"
+#include "py/gc.h"
+#include "py/stackctrl.h"
+#include "genhdr/mpversion.h"
+
+// This must be zero, we have no way to print.
+mp_uint_t mp_verbose_flag = 0;
+
+extern void mpycross_error(const char *buf, size_t len);
+
+STATIC void fd_print_vstr(void *env, const char *str, size_t len) {
+    vstr_t *out = (vstr_t *)env;
+    vstr_add_strn(out, str, len);
+}
+
+// No-op NLR. We accept a push (and do nothing). If anything attempts to jump,
+// fail immediately with no cleanup. The idea is that the host code running
+// the WASM binary will just handle the exception directly and then throw
+// away the entire execution environment.
+unsigned int nlr_push(nlr_buf_t *buf) {
+    (void)buf;
+    return 0;
+}
+
+NORETURN void nlr_jump(void *val) {
+    vstr_t buf;
+    vstr_init(&buf, 1024);
+    mp_print_t fd_print = {&buf, fd_print_vstr};
+    mp_obj_print_exception(&fd_print, MP_OBJ_FROM_PTR(val));
+
+    mpycross_error(buf.buf, buf.len);
+
+    // The host code must not attempt to resume the WASM execution after
+    // handling mpycross_error.
+    MP_UNREACHABLE
+}
+
+mp_import_stat_t mp_import_stat(const char *path) {
```

Comment:
Maybe it can disable `MICROPY_ENABLE_EXTERNAL_IMPORT` and then this function isn't needed (that could be done for the existing native variant as well).